accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,107 @@
+
+## Apple Silicon requires Signed Binaries
+
+Apple Silicon with Big Sur adds a new security policy which requires signed binaries by default. Unsigned binaries are simply killed without explanation. The user can expect an inscrutable message like `zsh killed: helloWorld` This directly affects our users who expect published binaries to just work.
+
+The developer would need to explicitly sign after publishing for Apple Silicon.  The sigining command is relatively simple `codesign -s <signature> <app-path>`.
+
+### Proposed changes
+
+This is largely TBD. The addition of a message during publish is the most likely design.
+
+## Universal Binaries
+
+As part of the architectural transition, Apple has switched the entire BigSur OS to use universal binaries. This means every file in the OS has both the `arm64` and the `x86_64` code on disk.
+
+Given both architectures, universal binaries are relatively easy to create. The general syntax would be `lipo <binaryArchA> <binaryArchB> -create -output <binaryUniversal>`
+
+### No planned universal binary .NET runtime
+
+At least for .NET 6, general universal binary support is not planned.
+
+General support would require:
+
+- All native binaries and libraries were built as universal binaries.
+- All managed libraries would contain IL which would work correctly on both platforms (no architecture specific `#if`)
+- All managed libraries contain either no `crossgen` native code, or `crossgen` code for both architectures.
+- The runtime would need to be taught how to select the correct architecture `crossgen` code.
+- The managed runtime file format would need to be extended to support multiple `crossgen` architectures.
+
+Given:
+
+- Universal binaries would effectively double the size of the runtime.
+- It requires significant investment
+
+This did not make .NET 6 investment plan.
+
+#### Shared dotnet host
+
+The impact to the shared `dotnet` host is discussed at length above. `dotnet` will be a universal binary.
+
+#### dotnet-sos and the SOS plugin
+
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. The plan would be to make the Apple Silcon SOS be a universal binary to allow SOS to work on `x86_64` and `arm64` processes.
+
+#### Apple Store requires universal binaries
+
+Publishing a .NET runtime to the Apple store is not a common customer scenario. However it is worth noting that the Apple store is requiring universal binaries.
+
+This in currently an unsupported scenario.
+
+It may be possible for a developer to generate a universal binary .NET 6 app, if each architecture is published as a single file app. Each architecture could be merged into a universal binary. This may just work, but it may also invalidate assumptions made by the single file app.
```

Comment:
I don't follow the distinction, did you have some kind of coreclr specific single-exe console in mind here?

Reply:
Yes.  One of the coreclr publishing types is internally called single-file